src/apps/vlan/vlan.lua

```diff
@@ -0,0 +1,174 @@
+module(..., package.seeall)
+
+local packet = require("core.packet")
+local bit = require("bit")
+local ffi = require("ffi")
+
+local C = ffi.C
+local receive, transmit = link.receive, link.transmit
+local cast = ffi.cast
+
+Tagger = {}
+Untagger = {}
+
+-- 802.1q
+local dotq_tpid = 0x8100
+local o_ethernet_ethertype = 12
+local uint32_ptr_t = ffi.typeof('uint32_t*')
+
+
+-- build a VLAN tag consisting of 2 bytes of TPID set to 0x8100 followed by the
+-- TCI field which in turns consists of PCP, DEI and VID (VLAN id). Both PCP
+-- and DEI is always 0
+local function build_tag(vid)
+   return ffi.C.htonl(bit.bor(bit.lshift(dotq_tpid, 16), vid))
+end
+
+-- pop a VLAN tag (4 byte of TPID and TCI) from a packet
+function pop_tag(pkt)
+   local payload = pkt.data + o_ethernet_ethertype
+   local length = pkt.length
+   pkt.length = length - 4
+   C.memmove(payload, payload + 4, length - o_ethernet_ethertype - 4)
+end
+
+-- push a VLAN tag onto a packet
+function push_tag(pkt, tag)
+   local payload = pkt.data + o_ethernet_ethertype
+   local length = pkt.length
+   pkt.length = length + 4
+   C.memmove(payload + 4, payload, length - o_ethernet_ethertype)
+   cast(uint32_ptr_t, payload)[0] = tag
+end
+
+-- extract TCI (2 bytes) from packet, no check is performed to verify that the
+-- packet is carrying a VLAN tag, if it's an untagged frame these bytes will be
+-- Ethernet payload
+function extract_tci(pkt)
+   return ffi.C.ntohs(ffi.cast("uint16_t*", packet.data(pkt) + o_ethernet_ethertype + 2)[0])
```

Comment:
Yes, I am aware. I explicitly did not want to require anything in lwaftr. I agree that these kinds of functions, if generally helpful, should be available in some core lib. I didn't do that here because a) it doesn't belong in same PR and b) I don't know if lwAFTR is the only program using this or if it would be adopted by more - maybe other devs prefer another pattern. Maybe I should have a local rd16 so it's easier to find afterwards if/when we introduce something into core!?

What do you mean with "cast is cached"? I don't know what to do with this information, do you mean I should change something? :)